src/librustc/middle/const_eval.rs

```diff
@@ -1103,10 +854,28 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
 
           let mut call_args = NodeMap();
           for (arg, arg_expr) in decl.inputs.iter().zip(args.iter()) {
+              let arg_hint = match ty_hint {
+                  ExprTypeChecked => ExprTypeChecked,
+                  UncheckedExprNoHint | UncheckedExprHint(_) => {
+                      if let Some(hint) = tcx.ast_ty_to_ty_cache.borrow().get(&arg.ty.id) {
```

Comment:
well, I guess you'd want a "fallible" version. But the truth is that reading the tables in this opportunistic way ("use the hint if the comipler happens to have produced it by now") is pretty dubious actually, and seems bound to fail when the programmer does simple things like rearrange the order of items. Maybe we should just remove these hints altogether if types are not available? I think probably so.